rust/src/fees.rs

```diff
@@ -0,0 +1,216 @@
+        },
+        DelegationCertificateEnum::PoolRetirement(_) => SMALL_ARRAY + LABEL_SIZE + UINT + HASH_OBJ,
+        DelegationCertificateEnum::GenesisKeyDeleg(_) => SMALL_ARRAY + 2 * LABEL_SIZE + HASH_OBJ,
+        DelegationCertificateEnum::MoveRewardsCert(mir) => {
+            SMALL_ARRAY + LABEL_SIZE + MAP_PREFIX + mir.move_instantaneous_reward.len() * (UINT + HASH_OBJ)
+        },
+    };
+    println!("cert_size({:?}) = {}", cert, x);
+    x
+}
+
+#[wasm_bindgen]
+pub fn txsize(tx: &TransactionBody, witness: &TransactionWitnessSet) -> usize {
+    // It seems that in Haskell we have Tx = {Body, vkeySigs, msigScripts, metadata}
+    // but in the CDDL we have Body as one then vkeySigs/msigScripts as another
+    // as the data is segwit.
+    // We might want to make our own version that wraps these then.
+    
+    // witness size
+    let signatures = witness.key_witnesses.as_ref().map_or(0, |keys| keys.len());
+    let script_nodes = witness.script_witnesses.as_ref().map_or(0, |scripts| scripts.0.iter().map(multisig_node_count).sum());
+    let witness_size = signatures * (ABSTRACT_SIZE_VKEY + ABSTRACT_SIZE_SIG)
+                     + HASH_OBJ * script_nodes
+                     + SMALL_ARRAY + LABEL_SIZE + MAP_PREFIX
+                     + HASH_OBJ * witness.script_witnesses.as_ref().map_or(0, Scripts::len);
+
+    let input_size = LABEL_SIZE + CBOR_TAG + ARRAY_PREFIX
+                   + tx.inputs.len() * (SMALL_ARRAY + UINT + HASH_OBJ);
+    let output_size = LABEL_SIZE + ARRAY_PREFIX
+                    + tx.outputs.len() * (SMALL_ARRAY + UINT + ADDRESS);
+    let certs_size = tx.certs.as_ref().map_or(0, |certs| certs.0.iter().map(cert_size).sum::<usize>());
+    let withdrawals_size = LABEL_SIZE + MAP_PREFIX
+                         + tx.withdrawals.as_ref().map_or(0, |wd| wd.len()) * (UINT + CREDENTIAL);
+    // TODO: updates
+    let update_size = ARRAY_PREFIX;
+    // TODO: metdata/hash
+    let metadata_size = ARRAY_PREFIX;
+    let metdata_hash_size = ARRAY_PREFIX;
+    
+    println!("input_size = {}", input_size);
+    println!("output_size = {}", output_size);
+    println!("certs_size = {}", certs_size);
+    println!("withdrawals_size = {}", withdrawals_size);
+    println!("FEE_SIZE = {}", FEE_SIZE);
+    println!("TTL_SIZE = {}", TTL_SIZE);
+    println!("update_size = {}", update_size);
+    println!("metdata_hash_size = {}", metdata_hash_size);
+    println!("witness_size = {}", witness_size);
+    println!("metadata_size = {}", metadata_size);
+    input_size + output_size + certs_size + withdrawals_size + FEE_SIZE + TTL_SIZE + update_size + metdata_hash_size + witness_size + metadata_size
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use js_chain_libs::*;
+    use address::*;
+
+    const HLEN: usize = 4;
+    const SLEN: usize = 13;
+
+    // TODO: compare raw bytes now that we have complete test vectors for this
```

Comment:
This is no longer relevant, right? Since we know `Hash` is the right size

Reply:
I was going to use it it to verify the general transaction structure against theres since it could be the same size but with some bad structure... Although I guess it would probably just lead to us realizing we're mixing definite/indefinite encodings somewhere and they wouldn't match byte-wise for that.